File: 3. Behavioral Design Patterns/Command/Typescript/CommandExample.ts
// Command Pattern Example in TypeScript
// Remote executes commands on Light and Stereo

interface Command { execute(): void; undo(): void }

class Light {
  on(){ console.log('Light ON') }
  off(){ console.log('Light OFF') }
}

class Stereo {
  on(){ console.log('Stereo ON') }
  off(){ console.log('Stereo OFF') }
}

class LightOnCommand implements Command {
  constructor(private l: Light){}
  execute(){ this.l.on() }
  undo(){ this.l.off() }
}

class LightOffCommand implements Command {
  constructor(private l: Light){}
  execute(){ this.l.off() }
  undo(){ this.l.on() }
}

class StereoOnCommand implements Command {
  constructor(private s: Stereo){}
  execute(){ this.s.on() }
  undo(){ this.s.off() }
}

class StereoOffCommand implements Command {
  constructor(private s: Stereo){}
  execute(){ this.s.off() }
  undo(){ this.s.on() }
}

class RemoteButton {
  constructor(private cmd: Command){}
  set(c: Command){ this.cmd = c }
  press(){ this.cmd.execute() }
  pressUndo(){ this.cmd.undo() }
}

;(function main(){
  const light = new Light();
  const stereo = new Stereo();
  const btn = new RemoteButton(new LightOnCommand(light));
  btn.press(); btn.pressUndo();
  btn.set(new StereoOnCommand(stereo));
  btn.press(); btn.pressUndo();
})()

export {}

File: 3. Behavioral Design Patterns/Command/Kotlin/CommandExample.kt
// Command Pattern Example in Kotlin
// Remote executes commands on Light and Stereo

interface Command { fun execute(); fun undo() }

class Light { fun on() = println("Light ON"); fun off() = println("Light OFF") }
class Stereo { fun on() = println("Stereo ON"); fun off() = println("Stereo OFF") }

class LightOnCommand(private val l: Light) : Command { override fun execute() = l.on(); override fun undo() = l.off() }
class LightOffCommand(private val l: Light) : Command { override fun execute() = l.off(); override fun undo() = l.on() }
class StereoOnCommand(private val s: Stereo) : Command { override fun execute() = s.on(); override fun undo() = s.off() }
class StereoOffCommand(private val s: Stereo) : Command { override fun execute() = s.off(); override fun undo() = s.on() }

class RemoteButton(private var cmd: Command) { fun set(c: Command){ cmd = c }; fun press() = cmd.execute(); fun pressUndo() = cmd.undo() }

fun main(){ val light = Light(); val stereo = Stereo(); val btn = RemoteButton(LightOnCommand(light)); btn.press(); btn.pressUndo(); btn.set(StereoOnCommand(stereo)); btn.press(); btn.pressUndo() }
